src/Microsoft.AspNet.Authentication/AutomaticAuthenticationHandler.cs

```diff
@@ -0,0 +1,98 @@
+    {
+        public virtual bool ShouldConvertChallengeToForbidden()
+        {
+            // Return 403 iff 401 and this handler's authenticate was called
+            // and the challenge is for the authentication type
+            return Response.StatusCode == 401 &&
+                AuthenticateCalled &&
+                ChallengeContext != null &&
+                ShouldHandleChallenge(ChallengeContext.AuthenticationSchemes);
+        }
+
+        protected async override Task InitializeCoreAsync()
+        {
+            if (Options.AutomaticAuthentication)
+            {
+                AuthenticationTicket ticket = await AuthenticateAsync();
+                if (ticket != null && ticket.Principal != null)
+                {
+                    SecurityHelper.AddUserPrincipal(Context, ticket.Principal);
+                }
+            }
+        }
+
+        public override void SignOut(ISignOutContext context)
+        {
+            // Empty or null auth scheme is allowed for automatic Authentication
+            if (Options.AutomaticAuthentication && string.IsNullOrWhiteSpace(context.AuthenticationScheme))
+            {
+                SignInContext = null;
+                SignOutContext = context;
+                context.Accept();
+            }
+
+            base.SignOut(context);
+        }
+
+        public override void Challenge(IChallengeContext context)
+        {
+            // Null or Empty scheme allowed for automatic authentication
+            if (Options.AutomaticAuthentication && 
+                (context.AuthenticationSchemes == null || !context.AuthenticationSchemes.Any()))
+            {
+                ChallengeContext = context;
+                context.Accept(BaseOptions.AuthenticationScheme, BaseOptions.Description.Dictionary);
+            }
+
+            base.Challenge(context);
+        }
+
+        /// <summary>
+        /// Automatic Authentication Handlers can handle empty authentication schemes
+        /// </summary>
+        /// <returns></returns>
+        public override bool ShouldHandleChallenge(IEnumerable<string> authenticationSchemes)
+        {
+            if (base.ShouldHandleChallenge(authenticationSchemes))
+            {
+                return true;
+            }
+
+            return Options.AutomaticAuthentication &&
+                (authenticationSchemes == null || !authenticationSchemes.Any());
+        }
+
+        /// <summary>
+        /// Override this method to deal with 401 challenge concerns, if an authentication scheme in question
+        /// deals an authentication interaction as part of it's request flow. (like adding a response header, or
+        /// changing the 401 result to 302 of a login page or external sign-in location.)
+        /// </summary>
+        /// <returns></returns>
+        protected override Task ApplyResponseChallengeAsync()
```

Comment:
Well its a bit confusing, cookies is overriding the sync version only, it would have to override the async version as well then, so this method is whats being used for cookies to handle the 401-403